ADC/Core/Src/main.c

```diff
@@ -0,0 +1,148 @@
+#include "init.h"
+#include "struct.h"
+
+ADC_Status	ADC_Config (uint32_t Channel);	//we call this function with the ADC channel we want to read
+void 		Wait (uint16_t);				//delay function, system clock based
+
+int main (void)
+{
+	uint16_t	measure 	= 0,	//ADC data read is stored here after a conversion
+				v_ref 		= 0;	//internal reference voltage measured (after calculation)
+	int16_t		temperature	= 0;	//temperature in Celsius degrees measured (after calculation)
+
+	ADC_Init();	//ADC initialization function
+
+	while(1)
+	{
+		ADC_Config(ADC_CHSELR_CHSEL18);		//configuring the ADC for internal temperature reading
+		ADC1->CR |= ADC_CR_ADSTART;			//starting the ADC
+		while(!(ADC1->ISR & ADC_ISR_EOC));	//and waiting for End Of Conversion flag to set
+		measure = ADC1->DR;					//storing the data read in measure
+
+		//temperature calculation as in RM example
+		temperature = ((measure * VDD_APPLI / VDD_CALIB) - (int32_t) *TEMP30_CAL_ADDR);
+		temperature = temperature * (int32_t)(130 - 30);
+		temperature = temperature / (int32_t)(*TEMP130_CAL_ADDR - *TEMP30_CAL_ADDR);
+		temperature = temperature + 30;			//the calculated temperature is now stored in temperature
+
+		ADC_Config(ADC_CHSELR_CHSEL17);		//configuring the ADC for internal reference voltage reading
+		ADC1->CR |= ADC_CR_ADSTART;			//starting the ADC
+		while(!(ADC1->ISR & ADC_ISR_EOC));	//and waiting for End Of Conversion flag to set
+		measure = ADC1->DR;					//storing the data read in measure
+
+		//Vref calculation as in RM
+		v_ref = (3 * (int32_t) (*VREFINT_CAL));
+		v_ref = (v_ref * 1000) / measure;		//*1000 so we have Vref in mV
+												//the calculated internal reference voltage is now stored in v_ref
+
+		asm("nop");		//so we can hold the uC here after reading the temperature and Vref
+	}
+
+	return 0;
+}
+
+ADC_Status ADC_Config (uint32_t Channel)
+{
+	if(ADC1->CR & ADC_CR_ADSTART)			//we have to be sure there's no ongoing conversion
+	{
+		ADC1->CR |= ADC_CR_ADSTP;			//if so, we have to set ADSTP bit
+		while(ADC1->CR & ADC_CR_ADSTP);		//and wait for it to clear
+	}
+	ADC1->CR |= ADC_CR_ADDIS;				//disables the ADC
+	while(ADC1->CR & ADC_CR_ADEN);
+	DISABLE_ALL_CHANNELS();					//and disables all channels
+
+	switch (Channel)
+	{
+		case ADC_CHSELR_CHSEL18:
+				ADC1->CHSELR |= ADC_CHSELR_CHSEL18;		//selecting the TSEN channel
+				ADC->CCR |= ADC_CCR_TSEN; 				//enables temperature sensor
+				Wait(TIME_10uSEC);						//we have to wait for the proper time for the Tsense to wake up
+		break;
+		case ADC_CHSELR_CHSEL17:
+				ADC1->CHSELR |= ADC_CHSELR_CHSEL17;		//selecting the VREF channel
+				ADC->CCR |= ADC_CCR_VREFEN; 			//enables internal reference voltage
+		break;
+		default:
+				return (ADC_ERROR);	//ADC error
+		break;
+	}
+
+	ADC1->CR |= ADC_CR_ADEN;				//then we enable the ADC
+	while(!(ADC1->ISR & ADC_ISR_ADRDY));	//and wait for it to be ready. (Can be handled by interrupt)
+
+	return 0;
+}
+
+void Wait (uint16_t time)
+{
+	while(time--);
+}
```

Comment:
Now this function is more efficient than the old one. Is 21 times enough? hahaha

Reply:
Yes, it is. I need a 10 uS delay, and 21 clock cycles are enough, plus there's the time spent to call the function, the while, etc

Reply:
don't forget to update the comment about the calculation of the 21 clk cycles

Reply:
well, I have these on main:
1)l. 17 void wait(uint16_t);	//delay function, system clock based
2)l. 72 wait(TIME_10uSEC);	//we have to wait for the proper time for the Tsense to wake up
and this one on macros.h:
3)l. 21 //--- TIME MANAGEMENT DEFINES
/*
 * Here we define constant values for Wait(uint16_t) function
 * System clock is running at 2.1 MHz
 * Ex.: 10 uS / (1 / 2.1 MHz) = 21 system clock cycles to wait for 10 uS
 */
#define TIME_10uSEC		21	//21 system clock cycles to wait for 10uS - time for the temperature sensor to wake up